lib/rex/ui/text/input/readline.rb

```diff
@@ -120,6 +120,21 @@ def intrinsic_shell?
       true
     end
 
+    @@in_shell=false
+    @@sess_cmd_count = 0
+
+    def self.sopen
+      @@in_shell=true
+    end
+
+    def self.sclose
+      @@in_shell=false
+    end
+ 
+    def in_shell?
+      @@in_shell
+    end
+
```

Comment:
Instead of adding class attributes and methods here to handle this, we should expand the history tracking capabilities in [`Rex::UI::Text::Shell`](https://github.com/rapid7/metasploit-framework/blob/80c3d82a9483d49b4181db828c7170db99080ddb/lib/rex/ui/text/shell.rb#L69-L73) to allow us to track multiple history files and switch between them at will. A final API from inside the command dispatcher that looks  like `use_history(Msf::Config.session_history_file)` would work pretty well, I think. Such an approach would also be able to solve #10149 by allowing us to switch history contexts before and after switching to IRB and pry.

Reply:
Asking coz I'm trying alternatives in which I push the data of other history files to the main one and not changing the file itself. I thought we cannot load or change the Msf::Config.histfile after loading the framework.

Reply:
We can switch the history file. In fact, that is what we want to do. We manage the history file ourselves in the class I linked (in fact, the file is managed per-shell and not framework-wide). `Msf::Config.histfile` isn't special, it's just where we decided to keep the path of the default history file so we could refer to it by a symbolic name instead of hard coding it everywhere. If we have it to where we use a different history file for `pry` and load it ourselves like we do for the main history file in the snippet I linked there should be no need to set new `pry` config options.

Reply:
Okay that info is really useful, thanks a lot. :+1: 
But I'm not able to override the existing `histfile` don't know why.
This method initializes meterpreter console. I am able to change prompt_char to `$` but `histfile` is still pointing to `Msf::Config.histfile`. Is there anything wrong in this?
```ruby
  def initialize(client)
    if (Rex::Compat.is_windows())
      super("meterpreter")
    else
      super("%undmeterpreter%clr", prompt_char = '$', histfile = Msf::Config.session_histfile) #new prompt_char and histfile for console.
    end
```
```
meterpreter $ irb
[*] Starting IRB shell...
[*] You are in the "client" (session) object

irb: warn: can't alias kill from irb_kill.
>> Msf::Config.session_histfile
=> "/home/pingport80/.msf4/session_history"
>> Msf::Config.history_file
=> "/home/pingport80/.msf4/history"
>> 
```

Reply:
```
if Readline::HISTORY.length == 0
```
I removed this condition from `Rex::Ui::Text::Shell` and now I'm able to don't know what this condition does.

Reply:
But yeah, I got it how it's loading the history file :+1:

Reply:
I can't see your current code, so I can't see what might be going wrong. Push it to this PR.

Reply:
okay sure

Reply:
[5a656fe](https://github.com/rapid7/metasploit-framework/pull/15062/commits/5a656fea682ce7782c4a80450af539b0303ff421). So these are the changes I did.

Reply:
@acammack-r7 , works now :)